src/daemon/rpmostreed-os.c

```diff
@@ -1788,78 +1788,58 @@ refresh_cached_update (RpmostreedOS *self, GError **error)
-  ot_repo = rpmostreed_sysroot_get_repo (rpmostreed_sysroot_get ());
+  OstreeSysroot *ot_sysroot = rpmostreed_sysroot_get_root (rpmostreed_sysroot_get ());
+  OstreeRepo *ot_repo = rpmostreed_sysroot_get_repo (rpmostreed_sysroot_get ());
 
-  booted = ostree_sysroot_get_booted_deployment (ot_sysroot);
-  if (booted && g_strcmp0 (ostree_deployment_get_osname (booted), name) == 0)
+  g_autofree gchar* booted_id = NULL;
+  OstreeDeployment *booted_deployment = ostree_sysroot_get_booted_deployment (ot_sysroot);
+  g_autoptr(GVariant) booted_variant = NULL; /* Strong ref as we reuse it below */
+  if (booted_deployment && g_strcmp0 (ostree_deployment_get_osname (booted_deployment), name) == 0)
     {
-      booted_variant = rpmostreed_deployment_generate_variant (ot_sysroot, booted, booted_id,
-                                                               ot_repo, error);
+      booted_variant =
+        g_variant_ref_sink (rpmostreed_deployment_generate_variant (ot_sysroot, booted_deployment, booted_id,
+                                                                    ot_repo, error));
       if (!booted_variant)
         return FALSE;
-      booted_id = rpmostreed_deployment_generate_id (booted);
+      booted_id = rpmostreed_deployment_generate_id (booted_deployment);
     }
-
-  deployments = ostree_sysroot_get_deployments (ot_sysroot);
-  for (guint i=0; i<deployments->len; i++)
-    {
-      if (g_strcmp0 (ostree_deployment_get_osname (deployments->pdata[i]), name) == 0)
-        {
-          default_variant = rpmostreed_deployment_generate_variant (ot_sysroot,
-                                                                    deployments->pdata[i],
-                                                                    booted_id,
-                                                                    ot_repo, error);
-          if (default_variant == NULL)
-            return FALSE;
-          break;
-        }
+  else
+    booted_variant = g_variant_ref_sink (rpmostreed_deployment_generate_blank_variant ());
+  rpmostree_os_set_booted_deployment (RPMOSTREE_OS (self), booted_variant);
+
+  g_autoptr(OstreeDeployment) rollback_deployment = NULL;
+  g_autoptr(OstreeDeployment) pending_deployment = NULL;
+  ostree_sysroot_query_deployments_for (ot_sysroot, name,
+                                        &pending_deployment, &rollback_deployment);
+
+  GVariant *default_variant = NULL; /* Floating */
+  if (pending_deployment)
+    {
+      default_variant = rpmostreed_deployment_generate_variant (ot_sysroot,
+                                                                pending_deployment,
+                                                                booted_id,
+                                                                ot_repo, error);
+      if (!default_variant)
+        return FALSE;
     }
+  else
+    default_variant = g_variant_ref (booted_variant); /* Default to booted */
```

Comment:
But we shouldn't `g_variant_ref` here if we want `default_variant` to remain an unowned binding, right?

Reply:
Yep good catch, fixed :arrow_down: